### PR TITLE
Proxito: fix https and canonical redirects

### PR DESCRIPTION
**Pull request by stsewd**
When doing a https redirect we were adding the default language and
version (/en/latest/) if the current url didn't have those components.
And similar when doing a redirect to a canonical domain.

Fix https://github.com/readthedocs/readthedocs.org/issues/8183

This is based on https://github.com/readthedocs/readthedocs.org/pull/8187 since we need to distinguish between redirects to subprojects and canonical redirects.

### CODE REVIEW

File: readthedocs/proxito/views/mixins.py [hunk 1]
```diff
@@ -169,19 +173,40 @@ def canonical_redirect(self, request, final_project, version_slug, filename):
         """
         Return a redirect to the canonical domain including scheme.
 
-        This is normally used HTTP -> HTTPS redirects or redirects to/from custom domains.
+        The following cases are covered:
+
+        - Redirect a custom from http to https (if supported)
```
Comment (ericholscher) [inline readthedocs/proxito/views/mixins.py:178]: `a custom`?

**Review by ericholscher — COMMENTED** [review]
I don't quite understand the issue. What's wrong with adding the full URL when redirecting users?

File: readthedocs/proxito/views/mixins.py [hunk 2]
```diff
@@ -169,19 +173,40 @@ def canonical_redirect(self, request, final_project, version_slug, filename):
...
+            to = resolve(
+                project=final_project,
+                version_slug=version_slug,
+                filename=filename,
+                query_params=parsed_from.query,
+                external=hasattr(request, 'external_domain'),
+            )
+            # When a canonical redirect is done, only change the domain.
+            if redirect_type == REDIRECT_CANONICAL_CNAME:
+                to_parsed = urlparse(to)
```
Comment (ericholscher) [inline readthedocs/proxito/views/mixins.py:203]: This should be called `parsed_to`, or the other should be `from_parsed` to keep them consistent.

**Review by ericholscher — APPROVED** [review]
This makes sense, though adds another redirect for a lot of URL's :/

File: readthedocs/proxito/constants.py [hunk 3]
```diff
@@ -0,0 +1,3 @@
+REDIRECT_HTTPS = 'https'
+REDIRECT_CANONICAL_CNAME = 'canonical-cname'
+REDIRECT_SUBPROJECT_MAIN_DOMAIN = 'subproject-main-domain'
```
Comment (ericholscher) [inline readthedocs/proxito/constants.py:3]: 💯

File: readthedocs/proxito/views/mixins.py (hunk 1 above)
Comment (stsewd) [reply readthedocs/proxito/views/mixins.py:178]: custom domain p:

Comment (stsewd) [reply readthedocs/proxito/views/mixins.py]: fixed

**Comment by stsewd** [comment]
This is based on https://github.com/readthedocs/readthedocs.org/pull/8187/, so I need a review on that one first.

**Pull request merged**